nflog/nflog.go

```diff
@@ -351,27 +408,37 @@ func (l *nlog) GC() (int, error) {
 
 // Query implements the Log interface.
 func (l *nlog) Query(params ...QueryParam) ([]*pb.Entry, error) {
-	q := &query{}
-	for _, p := range params {
-		if err := p(q); err != nil {
-			return nil, err
+	start := time.Now()
+	l.metrics.queriesTotal.Inc()
+
+	entries, err := func() ([]*pb.Entry, error) {
+		q := &query{}
+		for _, p := range params {
+			if err := p(q); err != nil {
+				return nil, err
+			}
+		}
+		// TODO(fabxc): For now our only query mode is the most recent entry for a
+		// receiver/group_key combination.
+		if q.recv == nil || q.groupKey == nil {
+			// TODO(fabxc): allow more complex queries in the future.
+			// How to enable pagination?
+			return nil, errors.New("no query parameters specified")
 		}
-	}
-	// TODO(fabxc): For now our only query mode is the most recent entry for a
-	// receiver/group_key combination.
-	if q.recv == nil || q.groupKey == nil {
-		// TODO(fabxc): allow more complex queries in the future.
-		// How to enable pagination?
-		return nil, errors.New("no query parameters specified")
-	}
 
-	l.mtx.RLock()
-	defer l.mtx.RUnlock()
+		l.mtx.RLock()
+		defer l.mtx.RUnlock()
```

Comment:
Typically I wouldn't try to avoid defer everywhere but in this case, since this is a method that is potentially called rather often it probably makes sense not to use it here. Plus there doesn't seem to be a possible error state within the 3 lines below, which would cause the mutex to stay locked. Please correct me if I'm wrong.